modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/AbstractSetOpNode.java

```diff
@@ -305,40 +314,29 @@ protected List<RowT> getOnMapper(int cnt) {
             return res;
         }
 
-        protected List<RowT> getOnSingleOrReducer(int cnt) {
+        private List<RowT> getResultRows(int cnt) {
             Iterator<Map.Entry<GroupKey, int[]>> it = groups.entrySet().iterator();
-
             List<RowT> res = new ArrayList<>(cnt);
 
             while (it.hasNext() && cnt > 0) {
                 Map.Entry<GroupKey, int[]> entry = it.next();
-
-                GroupKey key = entry.getKey();
-
-                Object[] fields = new Object[key.fieldsCount()];
-
-                for (int i = 0; i < fields.length; i++) {
-                    fields[i] = key.field(i);
-                }
-
-                RowT row = rowFactory.create(fields);
-
-                int[] cntrs = entry.getValue();
-
+                RowT row = createOutputRow(entry);
                 int availableRows = availableRows(entry.getValue());
 
+                updateAvailableRows(entry.getValue(), availableRows);
+
                 if (availableRows <= cnt) {
                     it.remove();
 
-                    if (availableRows == 0) {
-                        continue;
-                    }
-
                     cnt -= availableRows;
                 } else {
                     availableRows = cnt;
 
-                    decrementAvailableRows(cntrs, availableRows);
+                    assert availableRows > 0;
+                    assert all : "This branch should only be accessible for non distinct variant of a set operator";
+
+                    int[] cntrs = entry.getValue();
+                    cntrs[0] -= availableRows;
```

Comment:
position to decrement is an internal thing of particular Grouping implementation, thus I would prefer to hide it in `decrementAvailableRows` as it was before